core/src/OpenMPTarget/Kokkos_OpenMPTarget_Exec.hpp

```diff
@@ -1201,16 +1254,32 @@ KOKKOS_INLINE_FUNCTION void parallel_reduce(
     const Impl::TeamThreadRangeBoundariesStruct<
         iType, Impl::OpenMPTargetExecTeamMember>& loop_boundaries,
     const Lambda& lambda, const JoinType& join, ValueType& init_result) {
-  ValueType result = init_result;
+  ValueType* tmp_scratch =
+      static_cast<ValueType*>(loop_boundaries.team.impl_reduce_scratch());
 
-  for (iType i = loop_boundaries.start; i < loop_boundaries.end;
-       i += loop_boundaries.increment) {
-    ValueType tmp = ValueType();
-    lambda(i, tmp);
-    join(result, tmp);
+#pragma omp barrier
+  tmp_scratch[0] = init_result;
+#pragma omp barrier
+
+  if constexpr (std::is_arithmetic<ValueType>::value) {
+#pragma omp for reduction(+ : tmp_scratch[:1])
```

Comment:
this is wrong: it does + for the inter thread reduction, and then join within a thread.